### PR TITLE
Ghettochem

### PR DESCRIPTION
**Pull request by Kurfursten**
- Adds the Electrolyzer (plus construction)
- Adds the Centrifuge (plus construction)
- Adds the Mortar (plus construction)
- Sulphuric Acid can now be made with its real life formula. 2 S + 3 O2 + 2 H2O → 2 H2SO4
- Busta Nuts are now more salty, like Busta
- Bowls may now be made with 1 Wooden Plank
- Iron may now be obtained by grinding metal sheets

+Fully tested!

For more info on how Ghettochem tools work, check out this Wiki page: http://ss13.pomf.se/wiki/index.php/User:Kurfurst/Ghettochem

### CODE REVIEW

File: code/game/objects/items/weapons/grenades/chem_grenade.dm [hunk 1]
```diff
@@ -67,6 +67,11 @@
 		icon_state = initial(icon_state) +"_ass"
 		name = "unsecured grenade with [beakers.len] containers[detonator?" and detonator":""]"
 		stage = 1
+	else if(istype(W,/obj/item/stack/cable_coil/) && !beakers.len)
+		var/obj/item/stack/cable_coil/coil = W
+		coil.use(2)
```
Comment (clusterfack) [inline code/game/objects/items/weapons/grenades/chem_grenade.dm]: if(coil.amount < 2) return

File: code/game/objects/structures/watercloset.dm [hunk 2]
```diff
@@ -44,6 +45,16 @@
 	icon_state = "toilet[open][cistern]"
 
 /obj/structure/toilet/attackby(obj/item/I as obj, mob/living/user as mob)
+	if(open && cistern && !state && istype(I,/obj/item/stack/rods))
+		user << "<span class='notice'>You add the rods to the toilet.</span>"
+		var/obj/item/stack/rods/R = I
+		R.use(2)
```
Comment (Kurfursten) [inline code/game/objects/structures/watercloset.dm]: Needs a similar check for 2

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 3]
```diff
@@ -1381,3 +1382,315 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+/obj/item/weapon/electrolyzer/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(istype(W,/obj/item/weapon/wirecutters))
+		if(beakers.len)
+			user << "<span class='warning'> The electrolyzer contains beakers!</span>"
+			return
+		else
+			user << "<span class='notice'> You disassemble the electrolyzer.</span>"
+			var/turf/T = get_turf(src)
+			new /obj/item/stack/cable_coil(T)
+			new /obj/item/weapon/grenade/chem_grenade(T)
```
Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: Consider put in hand

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 4]
```diff
@@ -1381,3 +1382,315 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+
+/obj/item/weapon/electrolyzer/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(istype(W,/obj/item/weapon/wirecutters))
+		if(beakers.len)
+			user << "<span class='warning'> The electrolyzer contains beakers!</span>"
+			return
+		else
+			user << "<span class='notice'> You disassemble the electrolyzer.</span>"
+			var/turf/T = get_turf(src)
+			new /obj/item/stack/cable_coil(T)
```
Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: Sound add 2

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 5]
```diff
@@ -1381,3 +1382,315 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+
+/obj/item/weapon/electrolyzer/attack_self(mob/user as mob)
+	if(beakers.len)
+		for(var/obj/B in beakers)
+			if(istype(B))
+				beakers -= B
+				user.put_in_hands(B)
+
+/obj/item/weapon/electrolyzer/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(istype(W,/obj/item/weapon/wirecutters))
```
Comment (clusterfack) [inline code/modules/reagents/Chemistry-Machinery.dm]: you can use simple macros like iswirecutter(W), they're all in macros.dm

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 6]
```diff
@@ -1381,3 +1382,315 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+	var/list/juice_items = list (
+		/obj/item/weapon/reagent_containers/food/snacks/grown/tomato = list("tomatojuice" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/grown/carrot = list("carrotjuice" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/grown/berries = list("berryjuice" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/grown/banana = list("banana" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/grown/potato = list("potato" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/grown/lemon = list("lemonjuice" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/grown/orange = list("orangejuice" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/grown/lime = list("limejuice" = 0),
+		/obj/item/weapon/reagent_containers/food/snacks/watermelonslice = list("watermelonjuice" = 0),
```
Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: watermelonslice isn't a /grown/, so what's going to happen when the juicing part of the mortar asks for its potency?

Awful things, probably.

**Comment by Boxe** [comment]
>Sulphuric Acid can now be made with its real life formula. 2 S + 3 O2 + 2 H2O → 2 H2SO4

this line made me worry that you made it so that you had to make it manually rather than chem dispenser and from a quick read it seems i was wrong so thank you for not ruining everything.

**Comment by Iamgoofball** [comment]
:+1: creative use of Discount Dan's

**Comment by PJB3005** [comment]
:+1: Dis is amazing, now people can brew up their own drugs and sell them in the maint shafts.

File: code/modules/reagents/reagent_containers/food/snacks.dm [hunk 7]
```diff
@@ -1153,6 +1153,7 @@
 		..()
 		reagents.add_reagent("nutriment", 6)
 		reagents.add_reagent("bustanut", 6)
+		reagents.add_reagent("sodiumchloride", 6)
```
Comment (ComicIronic) [inline code/modules/reagents/reagent_containers/food/snacks.dm]: Why

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 8]
```diff
@@ -1414,3 +1415,315 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+			if (id[1] == 0) //Should always be true
+				if (crushable.reagents != null && crushable.reagents.has_reagent(id[1]))
+					reagents.add_reagent(id,min(crushable.reagents.get_reagent_amount(id), space))
+			else
+				reagents.add_reagent(id,min(id[1], space))
+		else if(istype(crushable, /obj/item/weapon/rocksliver)) //Xenoarch
+			var/obj/item/weapon/rocksliver/R = crushable
+			reagents.add_reagent(id,min(id[1], space), R.geological_data)
+		else
+			user << "<span class ='warning'> I dunno what this is.</span>"
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: Muh immulsions.

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 9]
```diff
@@ -1414,3 +1415,315 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+		new /obj/item/trash/bowl(user.loc)
+		return
+	if (crushable)
+		user << "<span class ='warning'> There's already something inside!</span>"
+		return 1
+	if (!is_type_in_list(O, blend_items) && !is_type_in_list(O, juice_items))
+		user << "<span class ='warning'> You can't grind that!</span>"
+		return ..()
+
+	user.before_take_item(O)
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: user.drop_item(src)
Why use before_take_item?

File: code/modules/reagents/reagent_containers/food/snacks.dm (hunk 7 above)
Comment (Kurfursten) [inline code/modules/reagents/reagent_containers/food/snacks.dm]: I wanted to provide greytide with more ready access to salt if the chef is being uncooperative.

This amount of salt should be harmless. I thought Busta's salty nature made it fitting.

File: code/modules/reagents/Chemistry-Machinery.dm (hunk 9 above)
Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: It (and the "updateUsrDialog" below) appears to be a typo related to SERIOUS HEAVY DUTY code stealing from the grinder.

**Comment by SkowronX** [comment]
In light of these new methods, you should also add make dispensers access restricted to scientists and chemists only.

**Comment by Iamgoofball** [comment]
@SkowronX no

ID locks a shit

**Comment by PJB3005** [comment]
goofball obviously doesn't observe the scientists during a regular round with miners.

File: code/game/objects/structures/watercloset.dm [hunk 10]
```diff
@@ -44,6 +45,17 @@
 	icon_state = "toilet[open][cistern]"
 
 /obj/structure/toilet/attackby(obj/item/I as obj, mob/living/user as mob)
+	if(open && cistern && !state && istype(I,/obj/item/stack/rods))
+		var/obj/item/stack/rods/R = I
+		if(R.amount < 2) return
+		user << "<span class='notice'>You add the rods to the toilet.</span>"
+		R.use(2)
+		state++
```
Comment (ririchiyo) [inline code/game/objects/structures/watercloset.dm]: I don't really like how states are used like this, could you change these to be defines so we can know exactly what state an object is in?

**Comment by Kurfursten** [comment]
DNM, one last bug to fix: need to make beakers not splash away contents when inserted into centrifuge (other containers like cans still work fine)

**Comment by Iamgoofball** [comment]
@Kurfursten add it to the whitelist in glass.dm

**Comment by Kurfursten** [comment]
@Iamgoofball With spectacular tunnel vision I actually already added the electrolyzer to that whitelist, but forgot the centrifuge.

**Comment by Iamgoofball** [comment]
10/10

File: code/game/objects/items/weapons/grenades/chem_grenade.dm [hunk 11]
```diff
@@ -67,6 +67,13 @@
 		icon_state = initial(icon_state) +"_ass"
 		name = "unsecured grenade with [beakers.len] containers[detonator?" and detonator":""]"
 		stage = 1
+	else if(istype(W,/obj/item/stack/cable_coil/) && !beakers.len)
+		var/obj/item/stack/cable_coil/coil = W
+		if(coil.amount < 2) return
+		coil.use(2)
+		var/obj/item/weapon/electrolyzer/E = new /obj/item/weapon/electrolyzer
+		user.before_take_item(src)
+		user.put_in_hands(E)
```
Comment (ComicIronic) [inline code/game/objects/items/weapons/grenades/chem_grenade.dm]: Gotta qdel(src) somewhere.

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 12]
```diff
@@ -1414,3 +1415,356 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+			beaker = null
+		user << "<span class='notice'>You remove everything from the centrifuge.</span>"
+	else
+		user << "<span class='warning'>There is nothing to eject!</span>"
+
+/obj/structure/centrifuge/verb/flush()
+	set name = "Flush"
+	set category = "Object"
+	set src in view(1)
+	usr << "<span class='notice'>The \the [src] groans as it spits out containers.</span>"
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: Just \The [src] works.

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 13]
```diff
@@ -1414,3 +1415,356 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+	name = "mortar"
+	desc = "This is a reinforced bowl, used for crushing reagents. Ooga booga Rockstop."
+	icon = 'icons/obj/food.dmi'
+	icon_state = "mortar"
+	flags = FPRINT  | OPENCONTAINER
+	volume = 50
+	amount_per_transfer_from_this = 5
+	//We want the all-in-one grinder audience
+
+	var/list/blend_items = list (
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: Did you copy these straight from the grinder?

Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: It's supposed to do the same stuff as the grinder. But no, I had to write it differently. Notice there are several key differences.

Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: Such as?

Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: 1. The formatting of the variables is different.
2. Plastellium isn't listed, as it was redundant.
3. The generic food snack option is formatted differently.
4. All of the values for special grinding results are directly in the code.

The grinder list flat-out won't work with my code, and visa versa.

File: code/modules/reagents/Chemistry-Machinery.dm (hunk 12 above)
Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: Should add a fingerprint when using this verb.

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 14]
```diff
@@ -1414,3 +1415,356 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+
+/obj/item/weapon/reagent_containers/mortar/attack_self(mob/user as mob)
+	if(!crushable)
+		user << "<span class='notice'>There is nothing to be crushed.</span>"
+		return
+	if (reagents.total_volume >= volume)
+		user << "<span class='warning'>There is no more space inside!</span>"
+		return
+	if(is_type_in_list(crushable, juice_items))
+		user << "<span class='notice'>You smash the contents into juice!</span>"
```
Comment (Kurfursten) [inline code/modules/reagents/Chemistry-Machinery.dm]: Consider adding a special branching option if in both lists, to allow you your choice of juice or dust.

**Pull request merged**